### ISSUE TITLE
RFC: Refactor (Unit) Test Infrastructure

### ISSUE THREAD

**Issue by phip1611** [cloud-hypervisor/cloud-hypervisor#5761]
TL;DR: I'd like to propose a (probably small?) refactoring of the (unit) test infrastructure so that `cargo test [--workspace --all-features]` actually performs something useful.

# Objective
A developer wants to easily run all (unit) tests without external dependencies locally.
One of the biggest contributions of Rust + Cargo and its ecosystem is not only the language but also the unified build system including the ability to run tests. `cargo test` is a handy way of executing unit tests without having to know anything about the specific project/repository. However, unfortunately `cargo test` isn't very useful in Cloud Hypervisor for the following reasons:
- many "unit" tests require a net capability
- they talk to a kernel module
- integration tests also reuse the _unit_ test infrastructure of Cargo
- CHV uses a "real" workspace and not a virtual workspace, which requires to add `--workspace` to run unit tests in the whole repo

To workaround these issues, there are a bunch of test-helper scripts in `./scripts`, such as `run_unit_tests.sh`. In CI, these scripts are wrapped by docker containers with special permissions (`-cap-add net_admin`). They create the special environment that is needed for those tests.

# Proposal
1. Unit tests should be unit tests and don't require special permissions or talk with the network. However, I fully understand benefiting from Rusts unit test infrastructure is a big convenience. A possible solution might be to add `#[cfg(test-network)]` and `#[cfg(test-integration)]` to all "unit" tests that need a special environment to be executed. The existing shell scripts can pass the corresponding `--cfg` flags to `rustc` to activate those tests.
2. By transforming CHV from a real workspace to a virtual workspace (TL;DR: move `./src`, `./tests`, and parts of `./Cargo.toml` to `./chv`), one can run `cargo build` and `cargo test` without the need for `--workspace`. I think this is much more intuitive. One can still limit the scope using `cargo test --package foo`.

In the end, the project becomes more of a "natural Cargo project" while still having the ability to have specialized test environments that are prepared with special helper scripts.

What do you think? If you are positive, I'm happy to work towards that goal.

Ping: @blitz @tpressure

**Comment by rbradford** [comment]
1. We used to have the integration tests gated under a feature - "integration_tests" - but it was removed in this PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3546. I'm sure @likebreath can elaborate further.

To be clear none of the unit tests actually require network access; some of the tests that involve creating TAP devices need `CAP_NET_ADMIN`. I wouldn't be opposed to feature gating the unit tests that require `CAP_NET_ADMIN`.

2. Big :-1:  for reorganising the source tree - it makes backporting and bug fixes and code archaeology much harder.

**Comment by likebreath** [comment]
`cargo test` was designed to support no only unit tests, but also integration tests (plus documentation tests) [1]. This is why we removed the feature (redundant) gates "integration_tests" a while ago. Following Cargo's folder structure, all unit tests are embedded in each crate, while integration tests are placed in the "tests" folder. To me, this is clear, simple, and intuitive.

Also, I concur the concern of reorganizing the source tree, particularly (assuming) the only benefit would be allow us not using "--workspace".

Please correct/educate me if I am missing anything, particularly the benefits of moving to the so-called "virtual workspace". Thank you.

[1] https://doc.rust-lang.org/rust-by-example/testing.html

**Comment by alyssais** [comment]
Is it possible to run just the unit tests?

**Comment by rbradford** [comment]
> Is it possible to run just the unit tests?

Yes, `scripts/run_unit_tests.sh`

**Comment by likebreath** [comment]
Close the issue for no activites.

**Comment by phip1611** [comment]
Sorry for the late reply, @likebreath . I was actually about to comment here soon again. I just wanted to wait until more people commented their opinion.

> Big 👎 for reorganising the source tree - it makes backporting and bug fixes and code archaeology much harder.

Fully agree, I forgot about backporting. With that in mind, a restructuring of the workspace would cause a lot of trouble.

> cargo test was designed to support no only unit tests, but also integration tests (plus documentation tests) [1]. This is why we removed the (redundant) feature gates "integration_tests" a while ago. [...]  https://doc.rust-lang.org/rust-by-example/testing.html

Thanks for the pointer to that document. To be honest, I never thought about the `tests` directory to be there for integration tests. I thought it just depends on the `taste` of the programmer where to put their "unit" tests. But that makes sense.

So I assume that `cargo test --tests` refers to everything inside `./tests`. The naming is a little unfortunate, I would have preferred/expected `cargo test --integration-tests` but this is out of scope of this repo.


> > Is it possible to run just the unit tests?

> Yes, scripts/run_unit_tests.sh

I still think needing a shell script for unit testing is at least counterintuitive for a typical cargo project, unfortunately.

> I wouldn't be opposed to feature gating the unit tests that require CAP_NET_ADMIN.

Sounds good! Maybe, this would make the testing story more convenient. I think, I might set up a PR eventually. However, I have to find some time as my focus in currently on other projects, unfortunately.



--
Thanks y'all for participating in the discussion!
